### ISSUE TITLE
Strange behavior for a Keras Code: Shows different behavior for two codes running by the exact same logic.

### ISSUE THREAD

**Issue by abhaskumarsinha** [keras-team/keras#16792]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 11 64 bit
- TensorFlow installed from (source or binary): Binary (Python PIP)
- TensorFlow version (use command below): v2.9.0-18-gd8ce9f9c301 2.9.1
- Python version: 3.9.5
- Bazel version (if compiling from source):
- GPU model and memory: AMD Radeon(TM) Graphics (AMD Ryzen 5 Processor over 8GB RAM)
- Exact command to reproduce: https://colab.research.google.com/drive/12bMfjZgld6ETzPhPsz7HzoWHdL8056yG?usp=sharing

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

So, I've two different codes for a DC GAN, which is basically a Generative Adverbial Neural Network using Cov Nets, despite similar model architecture and similar training steps, my code fails to converge in the way and speed of the second code.

Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or why the requested feature is needed.

My Notebook (First version): https://colab.research.google.com/drive/12bMfjZgld6ETzPhPsz7HzoWHdL8056yG?usp=sharing
GFG Version (Second Version): https://www.geeksforgeeks.org/deep-convolutional-gan-with-keras/

Results with my code vs the GFG Version of the Code.
GFG Version of Code:
[1]: https://i.stack.imgur.com/k2UVN.png
(GFG Version of the Code, after 1st Epoch and 128 batch size)

My version of Code:
[2]: https://i.stack.imgur.com/ciaAS.png

(My Version of the Code, after training 1000 batch size and 60 epochs in total)

As you can see, my model fails severely to converge to the desired result and for some reason, I notice mode collapse or something which doesn't lets my model converge to the desired results.

I'm posting the code of my training sequence as well as a bit of explaination of what it does in the following section:

```
def training_sequence(batch_size=1000, epochs=60):
    for epoch in range(epochs):
        print("Epoch: " + str(epoch + 1) + " of " + str(epochs))
        #Training Discriminator:

        GAN.layers[0].trainable = False
        GAN.layers[1].trainable = True

        #1 for real samples and 0 for fake samples
        fake_generated_examples = generator(generate_random_sequence(size = int(batch_size / 2)))
        zeros = tf.zeros((int(batch_size / 2), 1, 1, 1))

        #train discriminator
        discriminator.fit(fake_generated_examples, zeros)
        ones = tf.ones((int(batch_size / 2), 1, 1, 1))
        real_examples = x_train[int(batch_size / 2)*(epoch) : int(batch_size / 2)*(epoch + 1)].reshape(int(int(batch_size / 2)*(epoch + 1) - int(batch_size / 2)*(epoch)), 28, 28, 1)

        #train discriminator
        discriminator.fit(real_examples, ones)

        #Training Generator
        GAN.layers[0].trainable = True
        GAN.layers[1].trainable = False

        #train generator
        GAN.fit(generate_random_sequence(size = int(batch_size / 2)), ones)
```
There are two models (I'll post the summary just a bit after their introduction) -

Discriminator: Takes in the input of `(28, 28, 1)` and outputs:` [[[[1]]]]` for real samples (the ones from the dataset) and `[[[[0]]]]` for fake ones (which are generated from Generators.
GAN: It is a Sequence of Generator and Discriminator models. It takes in a random value of `(1, 100)` and returns the output from the discriminator.
`generate_random_sequences(size)` is just a function to generate a number of random 100 numbers of size rows and `(1,100)` columns which are fed into the generator.

`int(batch_size / 2)*(epoch)` used in the code is just used to select the first `batch_size` instances from the dataset in each epoch loop and keep selecting the next batch size and so on, so there is no repetition of selection of a data from dataset during each epoch and they remain unique.

Here's my model architecture:
[3]: https://i.stack.imgur.com/2XeyJ.png

As you can see it's very similar to GFG Version of Models.

The training logic is similar in both cases, we train discriminator and then, GAN by locking discriminator by discriminator.trainable = False and training Generator this way. Despite training steps being similar for both models and same architecture, my model fails severely to converge as like the first one. I suppose its a problem with me, but even if its the case of bug, any guide or help would be very appreciated.

**Describe the current behavior**.

The Model severely fails to converge, despite exactly same training steps with different techniques.

**Describe the expected behavior**.

The expected behavior is similar to the output image of GFG Version of code as shown above.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

Improving documentation to atleast address the problem of why, if its on my side.

**Standalone code to reproduce the issue**.

My Notebook: https://colab.research.google.com/drive/12bMfjZgld6ETzPhPsz7HzoWHdL8056yG?usp=sharing (BUG HERE!)
GFG Version:  https://www.geeksforgeeks.org/deep-convolutional-gan-with-keras/

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by tilakrayal** [comment]
@gadagashwini,
I was able to reproduce the issue on tensorflow [v2.8](https://colab.research.google.com/gist/tilakrayal/52f2c3e21533c1f1d2e4006d6984881c/2-8-gan.ipynb), [v2.9](https://colab.research.google.com/gist/tilakrayal/a5d8e4d0a024cf1cde866ccd0b30aeb0/2-9gan.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/0048af697246e04a1b1a9878439e0093/nightly-gan.ipynb). Kindly find the gist of it here.

**Comment by abhaskumarsinha** [comment]
Hello @tilakrayal @gadagashwini , I'd like to point out that I found the issue, which had crept in inside my code. The GFG version uses model.train_on_batch() for convergence whereas I had used model.fit() with batch_size as 1. So, apparently, that was the one causing the problems in initial convergence. Apologies for taking your time.

The initial convergence seems fine for now. Kindly close this ticket. In case, I find any later convergence issues, I'll ask for a reopening of this ticket. Thank you. Apologies.
